### PR TITLE
[WIP] libFirm method body code generation

### LINKED ISSUE #123: Symbol does not implement Hash by-ref
The `strtab::Symbol` type compares for `PartialEq` by-ref, but `Hash` is derived, so is by-value. This should be added. This is  a bug, since `sym1.hash() == sym2.hash()` does not imply that `sym1 == sym2`. Dunno why this never became a problem, we do use `HashMap`s with `Symbol` keys.

Maybe we can use `RefEquality` from #120 for `Symbol`s anyway

### PR DESCRIPTION
**Pull request by joshuabach**
cc #117 

Closes #123 

~Based on #134, that should be merged before this.~

This can allegedly generate all control-flow structures ~(execpt `while`)~. The if-diamond seems to be generated correctly, variable declaration (of ints) and initial assignments also look fine. 

~Expression handling is very limited. Currently only supported are integer literals and local variable / parameter references. This is currently blocked on getting type information from @hediet.~

# TODO
- [x] `this` parameter
- [x] `while`
- [x] Expressions

### CODE REVIEW

**Comment by problame** [comment]
Rebase this one on #137, use `git rebase -i 59e2c09a6fe6f5a4669a0b3d3049957ad53c016c` and drop phil's patches, they are already rebased in #137  ( https://github.com/comprakt/comprakt/pull/136/commits/9a558788f89935883fdf24d332bb16536209991f https://github.com/comprakt/comprakt/pull/136/commits/b2caa4984288b709c04a1408a869caa5355d3d61 and https://github.com/comprakt/comprakt/pull/136/commits/afcc2d7a124379e5c6e783e0613c1d5d63ff7917 )

File: compiler-lib/src/firm.rs [hunk 1]
```diff
@@ -8,75 +19,411 @@ pub struct Options {
...
+    ///
+    /// TODO non-raw-ptr abstraction for ret type; Box<dyn ValueNode> might
+    /// work, but unnecessary box
+    fn gen_expr(&mut self, expr: &Spanned<'src, ast::Expr<'src>>) -> *mut ir_node {
+        use self::ast::Expr::*;
+        match &**expr {
+            Int(lit) => {
+                let val = unsafe { new_tarval_from_long(lit.parse().unwrap(), mode::Is) };
+                self.graph.new_const(val).as_value_node()
+            }
```
Comment (flip1995) [inline compiler-lib/src/firm.rs]: `NegInt` is pretty much the same as `Int`, except
```rust
lit.parse().map_or_else(|| -2147483648, |v| -v)
```

File: helloworld.mj [hunk 2]
```diff
@@ -0,0 +1,9 @@
+class X {}
+class Main {
+    public int foo() {
+        int x = 32;
+        return 23;
+    }
+	public static void main(String[] args) {
```
Comment (reiner-dolp) [inline helloworld.mj]: mixed tabs and whitespaces? =)

**Comment by problame** [comment]
I created https://github.com/comprakt/comprakt/pull/new/problame/merge-hediet-typechecking-into-firm-method-bodies-and-start-using-it to experiment with #141 merged into this branch.
Maybe you can scavenge the merge commit + the refactoring I started.

However, I want this branch to stay the WIP-branch for the following 8 days
* => my branch should be considered an experiment that can be cherry-picked from
* => hedit/type-checking should be merged ASAP into this branch
  * maybe you can use my merge commit (but was small)

**Comment by reiner-dolp** [comment]
Marking all of phils reviews as resolved since they are heavily outdated. Comments that still apply should be reopend in another review.

**Comment by problame** [comment]
bors r=flip1995,reiner-dolp,problame <3

**Pull request merged**